src/Microsoft.ComponentDetection.Detectors/rust/RustCliDetector.cs

```diff
@@ -187,9 +203,81 @@ private static bool ShouldFallbackFromError(string error)
         return true;
     }
 
-    private static bool ParseDependency(string dependency, out string packageName, out string version, out string source)
+    private static bool ParseDependencyMetadata(string dependency, out string packageName, out string version, out string source)
     {
-        var match = DependencyFormatRegex.Match(dependency);
+        // There are a few different formats for pkgids: https://doc.rust-lang.org/cargo/commands/cargo-pkgid.html#description
+        // 1. name => packageName
+        // 2. name@version packageName@1.0.4
+        // 3. url => https://github.com/rust-lang/cargo
+        // 4. url#version => https://github.com/rust-lang/cargo#0.33.0
+        // 5. url#name => https://github.com/rust-lang/crates.io-index#packageName
+        // 6. url#name@version => https://github.com/rust-lang/cargo#crates-io@0.21.0
+
+        // First, try parsing using the old format in cases where a version of rust older than 1.77 is being used.
+        if (ParseDependencyCargoLock(dependency, out packageName, out version, out source))
+        {
+            if (!(string.IsNullOrEmpty(packageName) || string.IsNullOrEmpty(version)))
+            {
+                return true;
+            }
+        }
+
+        var match = DependencyFormatRegexPkgId.Match(dependency);
+        packageName = null;
+        version = null;
+        source = null;
+        if (!match.Success)
+        {
+            return false;
+        }
+
+        var firstGroup = match.Groups[1];
+        var secondGroup = match.Groups[2];
+        var thirdGroup = match.Groups[3];
+
+        // cases 3-6
+        if (Uri.IsWellFormedUriString(dependency, UriKind.Absolute))
+        {
+            // in this case, first group is guaranteed to be the source.
+            // packageName is also set here for case 3
+            source = firstGroup.Success ? firstGroup.Value : null;
+            packageName = source;
+
+            // if there is a third group, then the second must be packageName, third is version.
+            if (thirdGroup.Success)
+            {
+                packageName = secondGroup.Value;
+                version = thirdGroup.Value;
+            }
+
+            // if there is no third group, but there is a second, the second group could be either the name or the version, check if the value starts with a number (not allowed)
+            else if (secondGroup.Success)
```

Comment:
Can we try and get a test case covering these two below matches?

Reply:
changed the logic to use the pkgIds directly, so this whole parsing method is no longer needed.